### PR TITLE
Implement `size_hint` and `ExactSizedIterator` for `DecimalArray`

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?
Closes https://github.com/apache/arrow-rs/issues/1505

# Rationale for this change
 
I was doing some small cleanups in DataFusion https://github.com/apache/arrow-datafusion/pull/2107 and noticed that the DecimalArray iterators do not report their exact size. The exact size is used in several places for optimizations in Arrow

# What changes are included in this PR?
1. Implement size_hint and ExactSizedIterator for DecimalArray
2. Tests for same

# Are there any user-facing changes?

Can use DecimalArray::iter() in more situations

### CODE REVIEW

File: arrow/src/array/array_binary.rs [hunk 1]
```diff
@@ -1020,8 +1020,17 @@ impl<'a> std::iter::Iterator for DecimalIter<'a> {
...
     }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let remain = self.array.len() - self.current;
+        (remain, Some(remain))
+    }
 }
 
+/// iterator has have known size.
```
Comment (viirya) [inline arrow/src/array/array_binary.rs:1031]:
```suggestion
/// iterator has known size.
```

File: arrow/src/array/array_binary.rs [hunk 2]
```diff
@@ -1020,8 +1020,17 @@ impl<'a> std::iter::Iterator for DecimalIter<'a> {
...
+impl<'a> std::iter::ExactSizeIterator for DecimalIter<'a> {}
```
Comment (viirya) [inline arrow/src/array/array_binary.rs:1032]: Should we have it in `iterator.rs`? I see other `ExactSizeIterator`s implementations located there.

Comment (alamb) [reply arrow/src/array/array_binary.rs:1032]: 🤔  that is a good idea. I can certainly move it there.

**Review by viirya — APPROVED** [review]
this looks good to me.

**Comment by alamb** [comment]
Thanks for the review @viirya !

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
lgtm. thanks @alamb !

**Review by yjshen — APPROVED** [review]

**Pull request merged**